k2/python/tests/autograd_test.py

```diff
@@ -0,0 +1,294 @@
+#
+# See ../../../LICENSE for clarification regarding multiple authors
+
+# To run this single test, use
+#
+#  ctest --verbose -R autograd_test_py
+
+import unittest
+
+import k2
+import torch
+
+
+class TestGetTotScores(unittest.TestCase):
+
+    def test_tropical_single_fsa(self):
+        # best path arc indexes are: 1, 3, 5, 10
+        s = '''
+            0 4 1 1
+            0 1 1 1
+            1 2 1 2
+            1 3 1 3
+            2 7 1 4
+            3 7 1 5
+            4 6 1 2
+            4 8 1 3
+            5 9 -1 4
+            6 9 -1 3
+            7 9 -1 5
+            8 9 -1 6
+            9
+        '''
+        fsa = k2.Fsa.from_str(s)
+        fsa = k2.create_fsa_vec([fsa])
+        fsa.scores.requires_grad_(True)
+        log_like = k2.get_tot_scores(fsa,
+                                     log_semiring=False,
+                                     use_float_scores=True)
+
+        assert log_like == 14
+        assert log_like.dtype == torch.float32
+
+        log_like.sum().backward()
+        expected = torch.zeros(len(fsa.scores))
+        expected[torch.tensor([1, 3, 5, 10])] = 1
+        assert torch.allclose(fsa.scores.grad, expected)
+
+        # now for double
+        fsa.scores.grad = None
+        log_like = k2.get_tot_scores(fsa,
+                                     log_semiring=False,
+                                     use_float_scores=False)
+        assert log_like == 14
+        assert log_like.dtype == torch.float64
+        log_like.sum().backward()
+        assert torch.allclose(fsa.scores.grad, expected)
+
+    def test_tropical_multiple_fsas(self):
+        # best path:
+        #  states: 0 -> 1 -> 3 -> 7 -> 9
+        #  arcs:     1 -> 3 -> 5 -> 10
+        #  scores: 1 + 3 + 5 + 5 = 14
+        s1 = '''
+            0 4 1 1
+            0 1 1 1
+            1 2 1 2
+            1 3 1 3
+            2 7 1 4
+            3 7 1 5
+            4 6 1 2
+            4 8 1 3
+            5 9 -1 4
+            6 9 -1 3
+            7 9 -1 5
+            8 9 -1 6
+            9
+        '''
+
+        #  best path:
+        #   states: 0 -> 2 -> 3 -> 4 -> 5
+        #   arcs:     1 -> 4 -> 5 -> 7
+        #   scores: 6 + 4 + 3 + 0 = 13
+        s2 = '''
+            0 1 1 1
+            0 2 2 6
+            1 2 3 3
+            1 3 4 2
+            2 3 5 4
+            3 4 6 3
+            3 5 -1 2
+            4 5 -1 0
+            5
+        '''
+
+        #  best path:
+        #   states: 0 -> 2 -> 3
+        #   arcs:     1 -> 3
+        #   scores: 100 + 5.5 = 105.5
+        s3 = '''
+            0 1 1 10
+            0 2 2 100
+            1 3 -1 3.5
+            2 3 -1 5.5
+            3
+        '''
+
+        fsa1 = k2.Fsa.from_str(s1)
+        fsa2 = k2.Fsa.from_str(s2)
+        fsa3 = k2.Fsa.from_str(s3)
+
+        fsa1.scores.requires_grad_(True)
+        fsa2.scores.requires_grad_(True)
+        fsa3.scores.requires_grad_(True)
```

Comment:
seems we need to define a function to do something like `FsaVec.required_grad_(True)`?

Reply:
The scores are the only differentiable part of it, so if anything it would just be a wrapper doing fsa1.scores.requires_grad_(True)

Reply:
So do we need `requires_grad_` ?

Reply:
I think it's normal for that option to be supported.  I don't think we'll really be saving any work by having requires_grad_ be false, but also we need to propagate it appropriately as later operations may need it.  So: yes, we need it.

Reply:
Added.